src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStyle.cs

```diff
@@ -0,0 +1,367 @@
+          Optional<string> name = default(Optional<string>),
+          Optional<string> prefix = default(Optional<string>),
+          Optional<string> suffix = default(Optional<string>),
+          Optional<string> wordSeparator = default(Optional<string>),
+          Optional<Capitalization> capitalizationScheme = default(Optional<Capitalization>))
+        {
+            var newName = name.HasValue ? name.Value : this.Name;
+            var newPrefix = prefix.HasValue ? prefix.Value : this.Prefix;
+            var newSuffix = suffix.HasValue ? suffix.Value : this.Suffix;
+            var newWordSeparator = wordSeparator.HasValue ? wordSeparator.Value : this.WordSeparator;
+            var newCapitalizationScheme = capitalizationScheme.HasValue ? capitalizationScheme.Value : this.CapitalizationScheme;
+
+            if (newName == this.Name &&
+                newPrefix == this.Prefix &&
+                newSuffix == this.Suffix &&
+                newWordSeparator == this.WordSeparator &&
+                newCapitalizationScheme == this.CapitalizationScheme)
+            {
+                return this;
+            }
+
+            return new NamingStyle(this.ID,
+                newName, newPrefix, newSuffix, newWordSeparator, newCapitalizationScheme);
+        }
+
+        public string CreateName(IEnumerable<string> words)
+        {
+            var wordsWithCasing = ApplyCapitalization(words);
+            var combinedWordsWithCasing = string.Join(WordSeparator, wordsWithCasing);
+            return Prefix + combinedWordsWithCasing + Suffix;
+        }
+
+        private IEnumerable<string> ApplyCapitalization(IEnumerable<string> words)
+        {
+            switch (CapitalizationScheme)
+            {
+                case Capitalization.PascalCase:
+                    return words.Select(CapitalizeFirstLetter);
+                case Capitalization.CamelCase:
+                    return words.Take(1).Select(DecapitalizeFirstLetter).Concat(words.Skip(1).Select(CapitalizeFirstLetter));
+                case Capitalization.FirstUpper:
+                    return words.Take(1).Select(CapitalizeFirstLetter).Concat(words.Skip(1).Select(DecapitalizeFirstLetter));
+                case Capitalization.AllUpper:
+                    return words.Select(w => w.ToUpper());
+                case Capitalization.AllLower:
+                    return words.Select(w => w.ToLower());
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private string CapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToUpper(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        private string DecapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToLower(c)).Concat(chars.Skip(1)).ToArray());
```

Comment:
```C#
if (word.Length >= 1 && char.IsUpper(word[0]))
{
    var chars = word.ToCharArray();
    chars[0] = char.ToLower(chars[0]);
    return new string(chars);
}
else
{
    return word;
}
```

If it's available here, you could avoid allocating the char array altogether by using a ```StringBuilder``` from ```StringBuilderCache```.